plat/xen: Use -lgcc for Xen ARM64 builds with fp emulation

razvand opened the pull request:
Define `XEN_LINK_LIBGCC_FLAG` to link with `-lgcc`. This is required for routines for floating point emulation.

GitHub-Fixes: #1524

<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`arm64`]
 - Platform(s): [`xen`]
 - Application(s): [`nginx`]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

razvand committed:
plat/xen: Use -lgcc for Xen ARM64 builds

Define `XEN_LINK_LIBGCC_FLAG` to link with `-lgcc`. This is required for
routines for floating point emulation.

GitHub-Fixes: #1524

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

oleksiimoisieiev commented:
LGTM

Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

mogasergiu commented on plat/xen/Linker.uk:
```diff
@@ -1,5 +1,7 @@
 ifeq (x86_64,$(CONFIG_UK_ARCH))
 XEN_LDFLAGS-y += -Wl,-m,elf_x86_64
+else ifeq (arm64,$(CONFIG_UK_ARCH))
+XEN_LINK_LIBGCC_FLAG := -lgcc
```
I understand that you essentially adapted `KVM_LINK_LIBGCC_FLAG ` to Xen. However, just a question, is this really named accordingly in the first place (`KVM_LINK_LIBGCC_FLAG` included)? This is a flag that is directly passed to the linker. Shouldn't this have been `XEN_LDFLAGS-y += -lgcc`? (Same for the KVM one).

razvand replied:
@mogasergiu , the `KVM_LINK_LIBGCC_FLAG` was introduced and "touched" in three commits:

- commit [`eddaa9b24`](https://github.com/unikraft/unikraft/commit/eddaa9b24) by Jia He
- commit [`08a9ebf34`](https://github.com/unikraft/unikraft/commit/08a9ebf34) by @marcrittinghaus 
- commit [`46d98734c`](https://github.com/unikraft/unikraft/commit/46d98734c) by you

I didn't check to see why there is a separate variable defined (`KVM_LINK_LIBGCC_FLAG`), instead of simply adding it to the `KVM_FLAGS` variable.

If it would be better to add it to the `KVM_FLAGS` variable (seems OK to me)., I would approve and merge this PR and then create a separate PR that fixes both KVM and Xen.

mogasergiu replied:
Yes, you are right. The last two commits are just copy-paste though. It just randomly jumped in my face now, as a reviewer, that this name looks misaligned with the others.

> If it would be better to add it to the KVM_FLAGS variable (seems OK to me)., I would approve and merge this PR and then create a separate PR that fixes both KVM and Xen.

Sure, makes sense to me, if it makes sense to the other as well. That being said, I do not have any strong opinions regarding this, I was just merely pointing it out :D.

razvand replied:
@michpappas , as you're the assignee of this PR, what do you think? Is it OK if we kept it like this and I opened an issue and a pull request that moves the `-lgcc` flag from `KVM_LINK_LIBGCC_FLAG` to `KVM_FLAGS`?

michpappas replied:
I agree with @mogasergiu here, and especially on that it's good to have `-lgcc` appended to `LDFLAGS-y` in a separate line, as one can easily trace the rationale through the commit history. @razvand the change is tiny so it would be really quick to update this PR with an additional commit for KVM, but if you think otherwise we can also do as you suggest. So if you'd still rather do it later, please let me know and I'll proceed to trigger the merge.

razvand replied:
@michpappas , I updated the pull request with a new commit to have this fixed. Please check it out.

@mogasergiu , we cannot use `LDFLAGS`. These are linker option but NOT library options which have to go at the end of the link command. So I used `LDLIBS` as the linking option, based on the [implicit Make variable](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html).

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreistan26 reviewed (approved):
Tested with #1523, works great!

Reviewed-by: Andrei Stan [andreistan2003@gmail.com](mailto:andreistan2003@gmail.com)

razvand committed:
plat/kvm,xen: Use LDLIBS-y for -lgcc option

Use `KVM_LDLIBS-y` and `XEN_LDLIBS-y` for the `-lgcc` linker option.
Instead of the custom flags `KVM_LINK_LIBGCC_FLAG` and
`XEN_LINK_LIBGCC_FLAG`.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

michpappas commented on plat/xen/Linker.uk:
```diff
@@ -1,7 +1,8 @@
 XEN_LDFLAGS-y += -Wl,-m,elf_x86_64
 else ifeq (arm64,$(CONFIG_UK_ARCH))
-XEN_LINK_LIBGCC_FLAG := -lgcc
+KVM_LDFLAGS-y += -Wl,-m,aarch64elf
```
Looks like a typo, this should be `XEN_LDFLAGS-y`

michpappas reviewed (approved):
LGTM. If you wish, please shorten the commit message (e.g. `s/floating point emulation/fp emulation/`) to keep `checkpatch` happy.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
